### PR TITLE
z_message_staff (1 NON-MATCHING)

### PR DESCRIPTION
**Pull request by hensldm**
Starting to bring over the work in the z_message branch on Decompollaborate (https://github.com/Decompollaborate/mm/tree/z_message). This starts with the smallest file, z_message_staff which only has 3 functions. Unfortunately one has proven elusive.
A lot of the MessageContext struct documention comes from the wider message branch so should make more sense as more of message gets PRd.
Thanks to everyone who has so far contributed to the message branch :) .

### CODE REVIEW

**Review by AngheloAlf — APPROVED** [review]

File: include/z64save.h [hunk 1]
```diff
@@ -278,13 +278,7 @@ typedef struct SaveInfo {
     /* 0xEA8 */ u32 unk_EA8[2];                        // Related to blue warps
     /* 0xEB0 */ u32 stolenItems;                       // Items stolen by Takkuri and given to Curiosity Shop Man
     /* 0xEB4 */ u32 unk_EB4;
-    /* 0xEB8 */ u32 bankRupees;
-    /* 0xEBC */ u32 unk_EBC;
-    /* 0xEC0 */ u32 unk_EC0;                           // Fishing flags
-    /* 0xEC4 */ u32 unk_EC4;
-    /* 0xEC8 */ u32 horseBackBalloonHighScore;
-    /* 0xECC */ u32 lotteryCodeGuess;                  // Lottery code chosen by player (only uses lower three hex digits)
-    /* 0xED0 */ u32 shootingGalleryHighScores;         // High scores for both shooting galleries. Town uses lower 16 bits, Swamp uses higher 16 bits.
```
Comment (engineer124) [inline include/z64save.h:287]: I think these comments could be moved to the enum you made? Just to preserve this info

File: src/code/z_message.c [hunk 2]
```diff
@@ -110,7 +110,7 @@ s32 Message_ShouldAdvanceSilent(PlayState* play) {
     MessageContext* msgCtx = &play->msgCtx;
     Input* controller = CONTROLLER1(&play->state);
 
-    if (msgCtx->unk12020 == 0x10 || msgCtx->unk12020 == 0x11) {
+    if (msgCtx->textboxEndType == TEXTBOX_ENDTYPE_10 || msgCtx->textboxEndType == TEXTBOX_ENDTYPE_11) {
```
Comment (engineer124) [inline src/code/z_message.c:113]:
```suggestion
    if ((msgCtx->textboxEndType == TEXTBOX_ENDTYPE_10) || (msgCtx->textboxEndType == TEXTBOX_ENDTYPE_11)) {
```

File: src/code/z_message.c [hunk 3]
```diff
@@ -322,10 +322,10 @@ void Message_StartTextbox(PlayState* play, u16 textId, Actor* Actor) {
 
     msgCtx->ocarinaAction = 0xFFFF;
     func_80150D08(play, textId);
-    msgCtx->unkActor = Actor;
+    msgCtx->talkActor = Actor;
```
Comment (engineer124) [inline src/code/z_message.c:325]:
```suggestion
    msgCtx->talkActor = actor;
```
Not sure if this is caught in the main branch

File: src/code/z_message.c [hunk 4]
```diff
@@ -432,12 +432,12 @@ u32 func_80151C9C(PlayState* play) {
 #pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_80151DA4.s")
 
 void func_80152434(PlayState* play, u16 arg2) {
-    play->msgCtx.unk12046 = 0;
+    play->msgCtx.blockSunsSong = 0;
```
Comment (engineer124) [inline src/code/z_message.c:435]:
```suggestion
    play->msgCtx.blockSunsSong = false;
```
Could you make `blockSunsSong` bool?

File: src/code/z_message.c [hunk 5]
```diff
@@ -526,7 +527,7 @@ void func_80152EC0(PlayState* play) {
     if (1) {}
     if ((msgCtx->songPlayed < 0x17) && (msgCtx->songPlayed != 0xE) &&
         ((msgCtx->ocarinaAction < 0x43) || (msgCtx->ocarinaAction >= 0x47))) {
-        msgCtx->unk120B0 = 1;
+        msgCtx->ocarinaSongEffectActive = 1;
```
Comment (engineer124) [inline src/code/z_message.c:530]:
```suggestion
        msgCtx->ocarinaSongEffectActive = true;
```
Bool as well for this

File: src/code/z_message.c [hunk 6]
```diff
@@ -541,16 +542,16 @@ void func_80152EC0(PlayState* play) {
 
 #pragma GLOBAL_ASM("asm/non_matchings/code/z_message/func_80153750.s")
 
-void func_80153E7C(PlayState* play, void* arg1) {
-    if ((gSaveContext.options.language == 0) && (play->msgCtx.unk12090 == 0)) {
-        func_8014ADBC(play, arg1);
+void func_80153E7C(PlayState* play, Gfx** gfxP) {
+    if ((gSaveContext.options.language == 0) && (play->msgCtx.textIsCredits == 0)) {
```
Comment (engineer124) [inline src/code/z_message.c:546]:
```suggestion
    if ((gSaveContext.options.language == 0) && !play->msgCtx.textIsCredits) {
```

File: include/z64save.h [hunk 7]
```diff
@@ -510,10 +508,36 @@ typedef enum {
...
+    /* 3 */ HS_BOAT_ARCHERY,
+    /* 4 */ HS_HOSRE_BACK_BALLOON,
+    /* 5 */ HS_LOTTERY_GUESS,
+    /* 6 */ HS_SHOOTING_GALLERY
+} HighScore;
+
+#define HIGH_SCORE(type) (gSaveContext.save.saveInfo.highScores[(type)])
+
+#define HS_GET_BANK_RUPPEES() (HIGH_SCORE(HS_BANK_RUPPEES) & 0xFFFF)
+#define HS_SET_BANK_RUPPEES(ruppees) (HIGH_SCORE(HS_BANK_RUPPEES) = ((HIGH_SCORE(HS_BANK_RUPPEES) & 0xFFFF0000) | (ruppees)))
```
Comment (engineer124) [inline include/z64save.h:524]:
```suggestion
#define HS_GET_BANK_RUPEES() (HIGH_SCORE(HS_BANK_RUPEES) & 0xFFFF)
#define HS_SET_BANK_RUPEES(rupees) (HIGH_SCORE(HS_BANK_RUPEES) = ((HIGH_SCORE(HS_BANK_RUPEES) & 0xFFFF0000) | (rupees)))
```
Rupees has 1 p, not 2. Applies to a few places

File: src/overlays/actors/ovl_En_Kakasi/z_en_kakasi.c [hunk 8]
```diff
@@ -274,8 +274,9 @@ void func_8096FAAC(EnKakasi* this, PlayState* play) {
  * goes off every frame of song teach, but... doing what?
  */
 void func_8096FBB8(EnKakasi* this, PlayState* play) {
-    if ((play->msgCtx.unk12048 == 0) || (play->msgCtx.unk12048 == 1) || (play->msgCtx.unk12048 == 2) ||
-        (play->msgCtx.unk12048 == 3) || (play->msgCtx.unk12048 == 4)) {
+    if ((play->msgCtx.lastOcarinaButtonIndex == 0) || (play->msgCtx.lastOcarinaButtonIndex == 1) ||
+        (play->msgCtx.lastOcarinaButtonIndex == 2) || (play->msgCtx.lastOcarinaButtonIndex == 3) ||
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Kakasi/z_en_kakasi.c:278]:
```suggestion
    if ((play->msgCtx.lastOcarinaButtonIndex == OCARINA_BTN_A) || (play->msgCtx.lastOcarinaButtonIndex == OCARINA_BTN_C_DOWN) ||
        (play->msgCtx.lastOcarinaButtonIndex == 2) || (play->msgCtx.lastOcarinaButtonIndex == 3) ||
```
Could you fill out the enum for this now that it's being named?

File: src/overlays/actors/ovl_En_Kujiya/z_en_kujiya.c [hunk 9]
```diff
@@ -46,13 +46,13 @@ ActorInit En_Kujiya_InitVars = {
...
-#define CHECK_LOTTERY_NUMBERS                                                              \
-    (((u32)((void)0, gSaveContext.save.saveInfo.lotteryCodes[CURRENT_DAY - 1][0]) ==       \
-      ((((void)0, gSaveContext.save.saveInfo.lotteryCodeGuess & 0xFFFF) & 0xF00) >> 8)) && \
-     ((u32)((void)0, gSaveContext.save.saveInfo.lotteryCodes[CURRENT_DAY - 1][1]) ==       \
-      ((((void)0, gSaveContext.save.saveInfo.lotteryCodeGuess & 0xFFFF) & 0xF0) >> 4)) &&  \
-     ((u32)((void)0, gSaveContext.save.saveInfo.lotteryCodes[CURRENT_DAY - 1][2]) ==       \
-      (((void)0, gSaveContext.save.saveInfo.lotteryCodeGuess & 0xFFFF) & 0xF)))
+#define CHECK_LOTTERY_NUMBERS                                                        \
+    (((u32)((void)0, gSaveContext.save.saveInfo.lotteryCodes[CURRENT_DAY - 1][0]) == \
+      ((HS_GET_LOTTERY_CODE_GUESS() & 0xF00) >> 8)) &&                               \
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Kujiya/z_en_kujiya.c:51]:
```suggestion
#define CHECK_LOTTERY_NUMBERS()                                                        \
    (((u32)((void)0, gSaveContext.save.saveInfo.lotteryCodes[CURRENT_DAY - 1][0]) == \
      ((HS_GET_LOTTERY_CODE_GUESS() & 0xF00) >> 8)) &&                               \
```
Should this macro have empty params?

**Review by engineer124 — COMMENTED** [review]
Thanks for this!

File: src/code/z_message_staff.c [hunk 10]
```diff
@@ -17,12 +17,660 @@ void Message_FindCreditsMessage(PlayState* play, u16 textId) {
...
+
+            case 0x10: // MESSAGE_OCARINA
+                if ((i + 1) == msgCtx->textDrawPos) {
+                    Message_HandleOcarina(play);
+                    *gfxP = gfx;
+                    return;
+                }
+                break;
+
+            case 0xE: // MESSAGE_FADE
```
Comment (isghj5) [inline src/code/z_message_staff.c:278]: I'm curious why the enum is in the comment and the value is still literal, aren't they normally reversed?

Comment (hensldm) [reply src/code/z_message_staff.c:278]: Yeah, normally the `0xE` wouldn't be there. The main reason is that the `MESSAGE_FADE` probably won't be an enum, but a hex code define. It has to do with text extraction which we haven't done yet. If your curious I would recommend looking at `message_data_fmt.h` in OoT.

File: include/z64message.h [hunk 11]
```diff
@@ -82,103 +211,103 @@ typedef struct {
...
     /* 0x1203E */ s16 unk1203E;
-    /* 0x12040 */ struct Actor* unkActor;
+    /* 0x12040 */ struct Actor* talkActor;
     /* 0x12044 */ s16 unk12044;
-    /* 0x12046 */ s16 unk12046;
-    /* 0x12048 */ u8 unk12048; // EnKakasi
-    /* 0x12049 */ UNK_TYPE1 unk12049[0x1];
-    /* 0x1204A */ s16 unk1204A[0x5];
+    /* 0x12046 */ s16 blockSunsSong;
+    /* 0x12048 */ u8 lastOcarinaButtonIndex;
```
Comment (engineer124) [inline include/z64message.h:280]:
```suggestion
    /* 0x12048 */ u8 ocarinaButtonIndex;
```
Actually, something I've been wanting to do on the main message branch is to drop the `last` here. I think `ocarinaButtonIndex` should be enough?

File: include/z64save.h [hunk 12]
```diff
@@ -278,13 +278,7 @@ typedef struct SaveInfo {
...
     /* 0xEB0 */ u32 stolenItems;                       // Items stolen by Takkuri and given to Curiosity Shop Man
     /* 0xEB4 */ u32 unk_EB4;
-    /* 0xEB8 */ u32 bankRupees;
-    /* 0xEBC */ u32 unk_EBC;
-    /* 0xEC0 */ u32 unk_EC0;                           // Fishing flags
-    /* 0xEC4 */ u32 unk_EC4;
-    /* 0xEC8 */ u32 horseBackBalloonHighScore;
-    /* 0xECC */ u32 lotteryCodeGuess;                  // Lottery code chosen by player (only uses lower three hex digits)
-    /* 0xED0 */ u32 shootingGalleryHighScores;         // High scores for both shooting galleries. Town uses lower 16 bits, Swamp uses higher 16 bits.
+    /* 0xEB8 */ u32 highScores[7];
```
Comment (engineer124) [inline include/z64save.h:281]: Should this have a `MAX` in the HS enum used here?

File: src/code/z_message.c [hunk 13]
```diff
@@ -67,11 +67,11 @@ void func_80147520(void) {
...
     MessageContext* msgCtx = &play->msgCtx;
 
-    msgCtx->unk1204A[0] = 0xBD;
-    msgCtx->unk1204A[1] = 0xB8;
-    msgCtx->unk1204A[2] = 0xB3;
-    msgCtx->unk1204A[3] = 0xAE;
-    msgCtx->unk1204A[4] = 0xA9;
+    msgCtx->ocarinaButtonsPosY[0] = 0xBD;
+    msgCtx->ocarinaButtonsPosY[1] = 0xB8;
+    msgCtx->ocarinaButtonsPosY[2] = 0xB3;
```
Comment (engineer124) [inline src/code/z_message.c:72]:
```suggestion
    msgCtx->ocarinaButtonsPosY[OCARINA_BTN_A] = 0xBD;
    msgCtx->ocarinaButtonsPosY[OCARINA_BTN_C_DOWN] = 0xB8;
    msgCtx->ocarinaButtonsPosY[2] = 0xB3;
```
And for the rest. Can also probably be decimal, although we can do a decimal cleanup later

**Review by engineer124 — APPROVED** [review]

**Review by jpburnett — APPROVED** [review]

**Pull request merged**